tsfc/coffee_mode.py

```diff
@@ -0,0 +1,294 @@
+    :arg optimal_atomics: list of tuples of atomics to be used as common subexpression
+    :arg argument_indices: tuple of argument indices
+
+    :returns: a factorised :class:`MonomialSum` object, or the original object
+    if no changes are made
+    """
+    if not optimal_atomics or len(monomial_sum) < 2:
+        # Nothing to do
+        if isinstance(monomial_sum, MonomialSum):
+            return monomial_sum
+        else:
+            # input is list of monomials
+            new_monomial_sum = MonomialSum()
+            for m in monomial_sum:
+                new_monomial_sum.add(*m)
+            return new_monomial_sum
+    new_monomial_sum = MonomialSum()
+    # Group monomials with respect to each optimal atomic
+    factor_group = OrderedDict()
+    for monomial in monomial_sum:
+        for oa in optimal_atomics:
+            if oa in monomial.atomics:
+                # Add monomial to the list of corresponding optimal atomic
+                factor_group.setdefault(oa, []).append(monomial)
+                break
+        else:
+            # Add monomials that do no have argument factors to new MonomialSum
+            new_monomial_sum.add(*monomial)
+    # We should not drop monomials
+    assert sum(map(len, itervalues(factor_group))) + len(new_monomial_sum) == len(monomial_sum)
+
+    for oa, monomials in iteritems(factor_group):
+        if len(monomials) == 1:
+            # Just one monomial with this group, add to new MonomialSum straightaway
+            monomial, = monomials
+            new_monomial_sum.add(*monomial)
+            continue
+        all_atomics = []  # collect all atomics from monomials
+        all_rest = []  # collect all rest from monomials
+        for monomial in monomials:
+            _atomics = list(monomial.atomics)
+            _atomics.remove(oa)  # remove common factor
+            all_atomics.append(_atomics)
+            all_rest.append(monomial.rest)
+        # Create new MonomialSum for the factorised out terms
+        sub_monomial_sum = MonomialSum()
+        for _atomics, _rest in zip(all_atomics, all_rest):
+            sub_monomial_sum.add((), _atomics, _rest)
+        sub_monomial_sum = optimise_monomial_sum(sub_monomial_sum, argument_indices)
+        assert len(sub_monomial_sum) > 0
+        if len(sub_monomial_sum) == 1:
+            # result is a product, add to new MonomialSum directly
+            sub_monomial, = sub_monomial_sum
+            new_atomics = sub_monomial.atomics
+            new_atomics += (oa,)  # add back common factor
+            new_rest = sub_monomial.rest
+        else:
+            # result is a sum, we need to create new node
+            new_node = monomial_sum_to_expression(sub_monomial_sum)
+            new_atomics = [oa]
+            new_rest = one
+            if set(argument_indices) & set(new_node.free_indices):
+                new_atomics.append(new_node)
+            else:
+                new_rest = new_node
```

Comment:
This branching looks icky to me.  Is it because `monomial_sum_to_expression` does the wrong thing if there is only one monomial in the monomial sum?  If so, we should fix that.  It feels like the "generic" logic here should work for a "single monomial" case correctly too.